src/mscorlib/src/System/Threading/ThreadPool.cs

```diff
@@ -491,6 +491,9 @@ public IThreadPoolWorkItem Dequeue(ThreadPoolWorkQueueThreadLocals tl, ref bool
 
         internal static bool Dispatch()
         {
+            // Put threadpool thread on default context
+            ExecutionContextSwitcher.SetDefaults();
```

Comment:
With this:
> It seems like it may be better to keep the field set to null as far as possible rather than to keep it set to Default

I was thinking the EC field could be forced to null always instead of default, that is, it would never be set to `Default`, such that:
- `Capture` checks for null and returns `Default`
- `Run` would check for `Default` and set the field to null if not already null
- `Restore` would only check for null
- `ExecutionContextSwitcher.CurrentIsDefault` would only check for null

It seems like this would restore the number of branches and satisfy the same requirement that the write barrier is not called when running with the default context, maybe with some fewer loads of `Default`.

`Capture` has an odd contract, but its behavior is separate and can remain unchanged even if the EC field on `Thread` is kept as null instead of `Default`.

But this approach seems fine to me as well.

Regarding a one-time initialization, it would be possible to have the VM notify the managed side in `Thread.CurrentThread` if the `Thread` object was created, but since CurrentThread is called a lot, it may well be worse.